pkgs/development/beam-modules/build-mix.nix

```diff
@@ -0,0 +1,89 @@
+{ stdenv, writeText, elixir, erlang, hex, lib }:
+
+{ name
+, version
+, src
+, buildInputs ? [ ]
+, beamDeps ? [ ]
+, propagatedBuildInputs ? [ ]
+, postPatch ? ""
+, compilePorts ? false
+, meta ? { }
+, enableDebugInfo ? false
+, mixEnv ? "prod"
+, ...
+}@attrs:
+
+with lib;
+let
+  shell = drv: stdenv.mkDerivation {
+    name = "interactive-shell-${drv.name}";
+    buildInputs = [ drv ];
+  };
+
+  pkg = self: stdenv.mkDerivation (attrs // {
+    name = "${name}-${version}";
+    inherit version;
+    inherit src;
+
+    MIX_ENV = mixEnv;
+    MIX_DEBUG = if enableDebugInfo then 1 else 0;
+    HEX_OFFLINE = 1;
+
+    setupHook = attrs.setupHook or
+      writeText "setupHook.sh" ''
+      addToSearchPath ERL_LIBS "$1/lib/erlang/lib"
+    '';
+
+    buildInputs = buildInputs ++ [ elixir hex ];
+    propagatedBuildInputs = propagatedBuildInputs ++ beamDeps;
+
+    configurePhase = attrs.configurePhase or ''
+      runHook preConfigure
+
+      mkdir -p _build/$MIX_ENV/lib
+
+      while read -d ':' lib
+      do
+        for dir in $(ls $lib)
+        do
+          dest=$(echo "$dir" | cut -d '-' -f1)
+          [[ $MIX_DEBUG -eq 1 ]] && echo "Linking $lib/$dir to _build/$MIX_ENV/lib/$dest"
+          ln -s $lib/$dir _build/$MIX_ENV/lib/$dest
+        done
+      done <<< "$ERL_LIBS:"
+
+      runHook postConfigure
+    '';
+
+    buildPhase = attrs.buildPhase or ''
+      runHook preBuild
+      export HEX_HOME="$TEMPDIR/hex"
+      export MIX_HOME="$TEMPDIR/mix"
+      mix compile --no-deps-check
+      runHook postBuild
+    '';
+
+    installPhase = attrs.installPhase or ''
+      runHook preInstall
+      if [ -d "_build/shared" ]; then
+        MIX_ENV=shared
+      fi
+      mkdir -p "$out/lib/erlang/lib/${name}-${version}"
+      for reldir in src ebin priv include; do
```

Comment:
how about changing this to 
```
# the decision to choose /lib/erlang/lib as the out path is an arbitrary one by the nixpkgs beam maintainers
# there needs to be a uniformisation with the rebar3 output directory
mkdir -p "$out/lib/erlang/lib/${name}-${version}"

# some dependencies like db_connection will build into `shared`
for reldir in _build/{$MIX_ENV,shared}/lib/${name}/{src,ebin,priv,include} ; do
   
   if test -d $fd ; then
      # some dependencies will have symlinks, like phoenix priv directory
      cp  -Hrt "$out/lib/erlang/lib/${name}-${version}" "$fd"
   fi
done
```

so that we could remove the shared stuff above.

Reply:
I'm not sure what the `success=1` is about though

Reply:
> # some dependencies will have symlinks, like phoenix priv directory

I'm not sure what you mean by that. After building phoenix, I check everything in `/lib/erlang/lib/phoenix-1.5.8/priv/`. I'm not seeing any symlinks.

Reply:
The reason the symlink doesn't appear is because of the -H flag. I wanted to add the comment to explain why we use the -H flag here.
If you remove it, compilation will fail saying it can't find a file. if you look into the phoenix dependency you will see something like 
```
erlang/lib/phoenix-1.5.8🔒
❯ ls
Permissions Size User Date Modified Name
dr-xr-xr-x     - root  1 Jan  1970  ebin
lrwxr-xr-x    16 root  1 Jan  1970  priv -> ../../../../priv
```
Let me know if anything doesn't make sense.

Reply:
Thanks for explaining! I've added a code comment.